### PR TITLE
AVRO-2641: Fix for SpecificRecord String deserialization

### PR DESCRIPTION
**Pull request by mhelleborg**
Changed from casting object to String when schema expects String field. Since Avro can represent Strings in Java as either String, CharSequence or org.apache.avro.util.Utf8, the put can fail with a ClassCastException. By changing from cast to toString() the issue is circumvented.

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO-2641) 

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
 Updated existing unit tests to reflect changes.

### Commits

- [x] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

**Comment by mhelleborg** [comment]
@Fokko I agree it would be good to cover this with a regression test. I have not been able to reproduce it purely on the JVM though. The UTF8-issues might be because of different defaults in the .Net libraries. We are only seeing the issues with avro encoded on .Net and read out with Java.

**Comment by Fokko** [comment]
@blachniet By any chance, did you saw something similar with .Net?

**Comment by javierholguera** [comment]
We have had this issue with an all Java scenario:

1. Producer side - Kafka Connect using Debezium connector serializing with Avro 1.8.1.
2. Consumer side - Kafka Streams using Avro 1.9.1.

We worked around the issue using the templates shared in this PR

**Comment by gaisanshi** [comment]
We see this issue with 

- Producer side - Python 2.7 with avro 1.8.2
- Consumer side - Java consumer with Avro 1.9.1

File: lang/java/tools/src/test/compiler/output-string/avro/examples/baseball/Player.java [hunk 1]
```diff
@@ -117,8 +117,8 @@ public Player(java.lang.Integer number, java.lang.String first_name, java.lang.S
   public void put(int field$, java.lang.Object value$) {
     switch (field$) {
     case 0: number = (java.lang.Integer)value$; break;
-    case 1: first_name = (java.lang.String)value$; break;
-    case 2: last_name = (java.lang.String)value$; break;
+    case 1: first_name = value$!=null?value$.toString():null; break;
```
Comment (Fokko) [inline lang/java/tools/src/test/compiler/output-string/avro/examples/baseball/Player.java]:
```suggestion
    case 1: first_name = value$ != null ? value$.toString(): null; break;
```

File: lang/java/tools/src/test/compiler/output-string/avro/examples/baseball/Player.java [hunk 2]
```diff
@@ -117,8 +117,8 @@ public Player(java.lang.Integer number, java.lang.String first_name, java.lang.S
...
+    case 2: last_name = value$!=null?value$.toString():null; break;
```
Comment (Fokko) [inline lang/java/tools/src/test/compiler/output-string/avro/examples/baseball/Player.java]:
```suggestion
    case 2: last_name = value$ != null ? value$.toString(): null; break;
```

Comment (Fokko) [reply lang/java/tools/src/test/compiler/output-string/avro/examples/baseball/Player.java]: Thanks for picking this up so quickly, I think you've missed this one @mhelleborg

**Comment by RyanSkraba** [comment]
Hello, I agree, this PR would be welcome for 1.9.2.  The white space fixes are a bit more involved than the suggested edits (but pretty trivial).  I took the liberty of pushing to your branch, so this can be merged and cherry-picked.  Thanks!

**Pull request merged**